### ISSUE TITLE
rspfile_content cannot be empty

### ISSUE THREAD

**Issue by cdyson37** [ninja-build/ninja#1372]
Given the following `build.ninja`
````
rule test
  command = cat $out.rsp
  rspfile = $out.rsp
  rspfile_content = 
````
Running `ninja` gives

    ninja: error: build.ninja:5: rspfile and rspfile_content need to be both specified

To me, allowing `rspfile_content` to be non-empty is useful. I can workaround it by always putting something there, and always discarding the first word, but that's a bit hacky.

Incidentally https://github.com/ninja-build/ninja/blob/master/misc/ninja_syntax.py#L57 would need to change to `is not None` for this to be used from Python.

**Comment by nico** [comment]
You mean "allowing `rspfile_content` to be empty is useful" (?)

**Comment by cdyson37** [comment]
Yes - sorry! Tweaked the original report.

**Comment by nico** [comment]
After thinking about this, shouldn't your generator not use an rsp file at all if rspfile_content is "small" (unspecified what exactly this means, but a size of 0 qualifies as small :-) )? Ninja writes rspfiles in process, so it happens serially with everything, meaning that if you have many rspfiles it will slow down your build. So I'm not sure we want to make it convenient to write 0-sized rspfiles.

**Comment by cdyson37** [comment]
Thanks for thinking :-) In my case, the rspfile is (something like) a list of tests to run. This is in a very general setting, as I use this (meta) build system for multiple projects, so this could be
* empty
* small enough to fit on a commandline
* too big to fit on a commandline

It's awkward for me to have to special case things depending on the length of a list, as the tool being invoked then has to be invoked differently if an rspfile is used. As the documentation asserts, ninja is supposed to be easy for a machine to write, but forcing special cases makes it harder IMHO.

So for me, the only solution requiring me to write no special cases is to always use and rspfile, and always include a discarded first line so I don't fall foul of ninja's current behaviour forbidding an empty file.

> Ninja writes rspfiles in process, so it happens serially with everything

Not sure I understand the precise meaning of this: can other jobs run while an rspfile is being written for a particular job? It doesn't particularly affect me as I only have one or two rspfiles per project.

**Comment by nico** [comment]
The documentation says "when convenience and speed are in conflict, prefer speed". In general, I can't remember the documentation saying "easy to generate" (I looked again and couldn't see it, but it's possible I didn't look in the right place) – in fact, we have a saying "if the generator could do something, the generator should do it, not ninja".

For this particular issue, I think supporting empty rspfile_contents is ok. I'd imagine that it's a bit gnarly to implement since we currently use 'rspfile_contents is empty => it wasn't set" in the code and iirc the check for emptiness is a bit away from the reader, so we'd probably have to plumb some "rspfile_contents explicitly set but empty" bit through, or do something like "if rspfile_contents is set but empty, set it to ' '" at parse time. Neither sounds very appealing to me. If you can come up with a non-complicated patch for this I'm happy to merge it, but if it's gnarly I'm not sure I'd take it since it'd be complex code that exists solely for generators that use rsp files when they shouldn't.

> Not sure I understand the precise meaning of this: can other jobs run while an rspfile is being written for a particular job? It doesn't particularly affect me as I only have one or two rspfiles per project.

rspfiles can be written while child processes run. What I mean is the following. Ninja's main loop looks very roughly like:

```
while (work to do) {
    if (need to run more commands) {
      rspfile = write_rspfile_for_one_command_if_needed();
      launch_process_for_one_command(rspfile);
      continue;
   }
  if (a_command_is_done) {
    print_command_result();
    continue;
  }
  // if we get here, we still have work to do but can't, some command must have failed, error out
}
```
Say you're building a large project with many edges, and with some distcc like thing, so you can run many commands in parallel and build with `-j 200`. At the start of the build, ninja now starts 200 commands, which means it goes "write rspfile, start command, write rspfile, start command" if compiles use rspfiles. The commands run in the background fine, but the rspfile writing is in the critical path of starting commands. For a small project that isn't an issue, but for a small project most build systems will be fast.

**Comment by cdyson37** [comment]
Thanks again for looking at this. I'm not familiar with ninja's internals at all, but I'd have thought you could represent the rspfile contents as `NULL` for not set and `""` (i.e. valid pointer to a zero length string) for the empty string.

I think you're right: the documentation says that ninja files should be generated by another program, but it does not imply that it's supposed to be easy :stuck_out_tongue_closed_eyes:

> it'd be complex code that exists solely for generators that use rsp files when they shouldn't.

Turning it around: when _should_ you use an rsp file? Surely the answer is: when processing a list of items _might_ be too long for a commandline. Why add an additional _and is never empty_ to that?

Second bit: thanks for the explanation! Looks like to "prefer speed" the writing of the rspfile could in principle be hoisted into the parallel bit of the code? Maybe that's impractical.

**Comment by nico** [comment]
Doesn't the generator know the exact length of things it writes into rspfile_contents? If so, I'd say ideally you should use it if that's > 32k characters since that's the limit on Windows.

For the hoisting part, you'd have to create a process that then writes the rsp contents, and pipe the contents to that process, and then when it's done run the actual program (possibly as a child of the rsp writer process). That seems very complicated, I'm not sure if the additional process is faster than writing the file, and it's still strictly more work than not using rspfiles for common actions.

**Comment by cdyson37** [comment]
> Doesn't the generator know the exact length of things it writes into rspfile_contents?

Yes. But as I say, it's _sometimes_ less than the magic (platform-specific) limit, depending on e.g. the number of tests a particular project has. I don't want a special case for "the string is too long" any more than I want a special case for "the string is empty", so I always use an rspfile (one per build, so not terribly worried about performance).

**Comment by nico** [comment]
As I said, generators are free to make that choice, but then they produce ninja files that are suboptimal from a build perf point of view.
